src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -441,6 +456,206 @@ try
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'a', 'l', 't'), 1 }, // Contextual Alternates
+        { DWRITE_MAKE_FONT_FEATURE_TAG('l', 'i', 'g', 'a'), 1 }, // Standard Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'l', 'i', 'g'), 1 }, // Contextual Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('k', 'e', 'r', 'n'), 1 }, // Kerning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'a', 'r', 'k'), 1 }, // Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'k', 'm', 'k'), 1 }, // Mark to Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('d', 'i', 's', 't'), 1 } // Distances
+    };
+
+    // Update our feature map with the provided features
+    if (!features.empty())
+    {
+        for (const auto [tag, param] : features)
+        {
+            if (tag.length() == TAG_LENGTH)
+            {
+                featureMap.insert_or_assign(DWRITE_MAKE_FONT_FEATURE_TAG(til::at(tag, 0), til::at(tag, 1), til::at(tag, 2), til::at(tag, 3)), param);
+            }
+        }
+        _didUserSetFeatures = true;
+    }
+    else
+    {
+        _didUserSetFeatures = false;
+    }
+
+    // Convert the data to DWRITE_FONT_FEATURE and store it in a vector for CustomTextLayout
+    _featureVector.clear();
+    for (const auto [tag, param] : featureMap)
+    {
+        _featureVector.push_back(DWRITE_FONT_FEATURE{ tag, param });
+    }
+}
+
+// Routine Description:
+// - Updates our internal map of font axes with the given axes
+// - NOTE TO CALLER: Make sure to call _BuildFontRenderData after calling this for the axes changes
+//   to take place
+// Arguments:
+// - axes - the axes to update our map with
+void DxFontRenderData::_SetAxes(const std::unordered_map<std::wstring_view, int32_t>& axes)
+{
+    _axesVector.clear();
+
+    // Update our axis map with the provided axes
+#pragma warning(suppress : 26445) // the analyzer doesn't like reference to string_view
+    for (const auto& [axis, value] : axes)
+    {
+        if (axis.length() == TAG_LENGTH)
+        {
+            const auto dwriteTag = DWRITE_MAKE_FONT_AXIS_TAG(til::at(axis, 0), til::at(axis, 1), til::at(axis, 2), til::at(axis, 3));
+            _axesVector.emplace_back(DWRITE_FONT_AXIS_VALUE{ dwriteTag, gsl::narrow_cast<float>(value) });
+        }
+    }
+}
+
+// Method Description:
+// - Converts a DWRITE_FONT_STRETCH enum into the corresponding float value to
+//   create a DWRITE_FONT_AXIS_VALUE with
+// Arguments:
+// - fontStretch: the old DWRITE_FONT_STRETCH enum to be converted into an axis value
+// Return value:
+// - The float value corresponding to the passed in fontStretch
+float DxFontRenderData::FontStretchToWidthAxisValue(DWRITE_FONT_STRETCH fontStretch) noexcept
+{
+    if (gsl::narrow_cast<size_t>(fontStretch) > fontStretchEnumToVal.size())
+    {
+        fontStretch = DWRITE_FONT_STRETCH_NORMAL;
+    }
+
+    return til::at(fontStretchEnumToVal, fontStretch);

```

Comment:
i'm more in favor of putting the constant translation arrays inside these functions; it reads better and you don't need to mentally context-switch out of this function to figure out what it's doing.